In [24]:
import sqlite3
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib as plt
import sklearn

# Data Exploration

In [25]:
# Read sqlite query results into a pandas DataFrame
con = sqlite3.connect(r"C:\Users\maxso\PycharmProjects\ApartmentScraper\booli\scrapy_apartments.db")
df = pd.read_sql_query("SELECT * from apartment", con)


# Verify that result of SQL query is stored in the dataframe
print(df.head().T)

                                 0                 1  \
id                 /annons/3848254    /bostad/280424   
type                      Lägenhet          Lägenhet   
city                       Uppsala       Helsingborg   
street                   Torngatan  Lilla Bergaliden   
street_num                      24                 1   
district              Kapellgärdet           Centrum   
postal_code                  75423             25223   
latitude                   59.8664           56.0449   
longitude                  17.6387           12.7005   
rooms                            2                 2   
square_meters                   55                74   
floor                            1                 2   
housing_society       BRF Tornet 1    BRF Minerva 16   
construction_year             2011              1939   
fee                           3126              5028   
price                    2.095e+06         2.175e+06   

                                               

In [26]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6911 entries, 0 to 6910
Data columns (total 16 columns):
id                   6911 non-null object
type                 6911 non-null object
city                 6911 non-null object
street               6805 non-null object
street_num           6805 non-null object
district             6911 non-null object
postal_code          6911 non-null int64
latitude             6911 non-null float64
longitude            6911 non-null float64
rooms                6911 non-null float64
square_meters        6911 non-null int64
floor                6911 non-null int64
housing_society      5909 non-null object
construction_year    6096 non-null float64
fee                  6086 non-null float64
price                6578 non-null float64
dtypes: float64(6), int64(3), object(7)
memory usage: 864.0+ KB


In [27]:
print(df.describe())

        postal_code     latitude    longitude        rooms  square_meters  \
count   6911.000000  6911.000000  6911.000000  6911.000000    6911.000000   
mean   34252.778469    58.438192    15.793085     2.742295      72.000434   
std    23931.217763     1.453051     2.498384     1.507556      41.477190   
min    11121.000000    55.510615    11.710321     1.000000       6.000000   
25%    12473.000000    57.699510    12.982253     2.000000      44.000000   
50%    21633.000000    59.294872    17.617292     2.000000      62.000000   
75%    58232.000000    59.348830    18.013328     3.000000      87.000000   
max    75758.000000    60.146462    18.391504    18.000000     430.000000   

            floor  construction_year           fee         price  
count  6911.00000        6096.000000   6086.000000  6.578000e+03  
mean      2.47012        1968.389928   3545.184850  3.423673e+06  
std       2.72963          39.190492   1414.889842  3.086944e+06  
min       0.00000        1643.000000  

In [28]:
df.isnull().sum()

id                      0
type                    0
city                    0
street                106
street_num            106
district                0
postal_code             0
latitude                0
longitude               0
rooms                   0
square_meters           0
floor                   0
housing_society      1002
construction_year     815
fee                   825
price                 333
dtype: int64

In [29]:
df.isnull().sum()/df.count() * 100

id                    0.000000
type                  0.000000
city                  0.000000
street                1.557678
street_num            1.557678
district              0.000000
postal_code           0.000000
latitude              0.000000
longitude             0.000000
rooms                 0.000000
square_meters         0.000000
floor                 0.000000
housing_society      16.957184
construction_year    13.369423
fee                  13.555702
price                 5.062329
dtype: float64

we can see that around 5%, or 333 rows have missing values for the target variable. We can take these aside at a later point and, given that we are successful in creating an adequate model, use this model to predict these values. These rows can then be used to re-train a new model on. 

We can also see that almost 1.6% of *street* and *street_num* are missing. However, since *postal_code* technically contains roughly the same information, this will probably not be cause of any concern. We might even consider dropping these rows entirely.

Around 17% of the *housing_society* values are missing. This is what we will focus on exploring next.

## Housing Society exploration

In [30]:
df[df.housing_society.isnull()].head(50)

id        type         city                 street  \
6    /bostad/1846982       Villa  Helsingborg         Rusthållsgatan   
7    /bostad/2247046       Villa  Helsingborg            Grännagatan   
10   /bostad/2242476       Villa  Helsingborg                Alvägen   
33   /annons/3887215       Villa     Västerås         Fältspatsvägen   
36   /bostad/2073765       Villa     Västerås           Malörtsvägen   
37   /annons/3891158      Parhus  Helsingborg               Rååvägen   
51   /bostad/2815284       Villa  Helsingborg              Långgatan   
52   /annons/3877132       Villa  Helsingborg              Lövstigen   
53   /annons/3877528       Villa  Helsingborg         Färjemansgatan   
68   /bostad/2805939       Villa  Helsingborg          Brännerigatan   
76   /annons/3847844       Villa  Helsingborg  Övre Långvinkelsgatan   
79   /bostad/1827908       Villa  Helsingborg        Prästkragegatan   
87   /bostad/1848983      Radhus     Västerås           Råbocksvägen   
89   /annons/3441795  Fritidshus     Västerås         Lönnbacksvägen   
91   /bostad/2090288       Villa     Västerås              Tallvägen   
93   /annons/3645502    Lägenhet     Västerås           Notuddsallén   
100  /bostad/2989169       Villa     Västerås           Rektorsgatan   
103  /annons/3773854       Villa     Västerås          Sevalla Väsby   
107  /annons/3782832    Lägenhet     Västerås          Poseidongatan   
108  /bostad/4067860      Radhus     Västerås          Fåfänganvägen   
110  /annons/3648513       Villa     Västerås               Östanbro   
111  /annons/3648523       Villa     Västerås               Östanbro   
113  /bostad/1853113       Villa  Helsingborg          Röamöllagatan   
116  /annons/3448556        Gård  Helsingborg     Södra Bokhagsvägen   
118  /annons/3843289       Villa  Helsingborg      Johan Banérs gata   
129  /annons/3821669      Radhus  Helsingborg          Saturnusgatan   
130  /bostad/2359982       Villa  Helsingborg            Kruareliden   
132  /annons/3609756       Villa  Helsingborg                   None   
134  /annons/3825807  Fritidshus  Helsingborg          Brytstugan Nr   
140  /annons/3770806      Radhus  Helsingborg                   None   
150  /annons/3692060       Villa  Helsingborg      Johan Banérs gata   
151  /bostad/2795654       Villa  Helsingborg          Traktörsgatan   
160  /annons/3352940    Kedjehus  Helsingborg           Hyacintgatan   
170  /bostad/1520486       Villa  Helsingborg             Briggvägen   
175  /bostad/1535383        Gård  Helsingborg        Bältebergavägen   
184  /bostad/2127857       Villa  Helsingborg         Jonstorpsvägen   
185  /bostad/2101420       Villa     Västerås        Lykttändargränd   
187  /bostad/3013086  Fritidshus     Västerås         Bergskogsvägen   
191  /annons/3826156       Villa     Västerås                Gottsta   
197  /annons/3838530       Villa     Västerås           Boskapsvägen   
198  /annons/3838544      Parhus     Västerås           Boskapsvägen   
201  /annons/3734183    Lägenhet     Västerås            Stora gatan   
215  /annons/3654849       Villa  Helsingborg           Höganäsgatan   
217  /bostad/1674775       Villa  Helsingborg         Jordugglegatan   
220  /bostad/4066096      Radhus  Helsingborg           Vildrosgatan   
223  /annons/3873501       Villa  Helsingborg        Kvarnstensgatan   
230  /annons/3875501      Radhus  Helsingborg          Bjälloxegatan   
232  /bostad/1989214       Villa     Västerås            Vargbogatan   
237  /bostad/1711537       Villa  Helsingborg        Kungsfågelgatan   
242  /bostad/1771609      Parhus  Helsingborg          Missionsgatan   

    street_num                 district  postal_code   latitude  longitude  \
6           5A                  Miatorp        25231  56.015764  12.725562   
7            7              Fältabacken        25250  56.039595  12.727825   
10          13                    Laröd        25482  56.093354  12.647752   
33           1                   

It looks like it's mostly different kinds of houses such as villas and terrace houses that are missing *house_society*, which is not very surprising. It is mainly *Lägenheter*/apartments that are tied to housing societies.

In [31]:
df.groupby("type").count()

id  city  street  street_num  district  postal_code  latitude  \
type                                                                          
Fritidshus   104   104     101         101       104          104       104   
Gård           9     9       8           8         9            9         9   
Kedjehus      40    40      39          39        40           40        40   
Lägenhet    5990  5990    5913        5913      5990         5990      5990   
Parhus        53    53      49          49        53           53        53   
Radhus       191   191     187         187       191          191       191   
Villa        524   524     508         508       524          524       524   

            longitude  rooms  square_meters  floor  housing_society  \
type                                                                  
Fritidshus        104    104            104    104                1   
Gård                9      9              9      9                0   
Kedjehus           40     40             40     40                4   
Lägenhet         5990   5990           5990   5990             5817   
Parhus             53     53             53     53               24   
Radhus            191    191            191    191               61   
Villa             524    524            524    524                2   

            construction_year   fee  price  
type                                        
Fritidshus                 82     3    104  
Gård                        9     0      8  
Kedjehus                   39     5     38  
Lägenhet                 5245  5957   5677  
Parhus                     46    33     49  
Radhus                    186    85    189  
Villa                     489     3    513

In [32]:
#num of null-values per type
df.housing_society.isnull().groupby(df["type"]).sum()

type
Fritidshus    103.0
Gård            9.0
Kedjehus       36.0
Lägenhet      173.0
Parhus         29.0
Radhus        130.0
Villa         522.0
Name: housing_society, dtype: float64

In [33]:
#percentage of null-values per type
df.housing_society.isnull().groupby(df["type"]).mean() * 100

type
Fritidshus     99.038462
Gård          100.000000
Kedjehus       90.000000
Lägenhet        2.888147
Parhus         54.716981
Radhus         68.062827
Villa          99.618321
Name: housing_society, dtype: float64

In [34]:
df[(df["type"] == 'Lägenhet') & (df["housing_society"].isnull())]

id      type       city                             street  \
93    /annons/3645502  Lägenhet   Västerås                       Notuddsallén   
107   /annons/3782832  Lägenhet   Västerås                      Poseidongatan   
201   /annons/3734183  Lägenhet   Västerås                        Stora gatan   
295   /annons/3893744  Lägenhet   Västerås  Välplanerad Etta - Stentorpsgatan   
334   /annons/3785800  Lägenhet    Uppsala                               None   
464   /annons/3851345  Lägenhet    Uppsala                     Wivalliusgatan   
474    /bostad/224045  Lägenhet    Uppsala                  Petterslundsgatan   
478   /annons/3846058  Lägenhet    Uppsala                      Åskmolnsvägen   
479   /annons/3846064  Lägenhet    Uppsala                               None   
480   /annons/3846073  Lägenhet    Uppsala                               None   
511   /annons/3817658  Lägenhet    Uppsala                               None   
538   /annons/3768315  Lägenhet    Uppsala                               None   
605   /annons/3744257  Lägenhet    Uppsala                        Knutbyvägen   
611   /annons/3771797  Lägenhet    Uppsala                               None   
639   /annons/3827274  Lägenhet    Uppsala                               None   
640   /annons/3827282  Lägenhet    Uppsala                               None   
690   /annons/3875449  Lägenhet    Uppsala                               None   
829   /annons/3862352  Lägenhet     Örebro                               None   
836   /annons/3859927  Lägenhet     Örebro                        Ekängsgatan   
844   /annons/3843381  Lägenhet     Örebro               Södra Grevrosengatan   
931   /annons/3865771  Lägenhet     Örebro                               None   
932   /annons/3865788  Lägenhet     Örebro                               None   
933   /annons/3865802  Lägenhet     Örebro                               None   
934   /annons/3865806  Lägenhet     Örebro                               None   
983   /annons/3866436  Lägenhet  Linköping                        Ulvåsavägen   
993   /annons/3889986  Lägenhet  Linköping                Hugo Theorells gata   
994   /annons/3881690  Lägenhet  Linköping                          Sivs gata   
1022  /annons/3853993  Lägenhet  Linköping                      Panghus Etapp   
1028  /annons/3854641  Lägenhet  Linköping                        Ekängsvägen   
1030  /annons/3827851  Lägenhet  Linköping                         Sisalgränd   
...               ...       ...        ...                                ...   
4049  /annons/3900392  Lägenhet    Uppsala                      Altfiolsvägen   
4183  /annons/3913717  Lägenhet    Uppsala                          Lings väg   
4205  /annons/3877808  Lägenhet   Västerås                       Notuddsallén   
4222  /annons/3922704  Lägenhet   Göteborg                    Lindholmshamnen   
4313  /annons/3833353  Lägenhet   Västerås                     Smörblomsgatan   
4630  /annons/3924566  Lägenhet      Malmö                               None   
4732  /annons/3920451  Lägenhet      Malmö                         Kapellplan   
4739  /annons/3916605  Lägenhet      Malmö                               None   
4745  /annons/3916840  Lägenhet      Malmö                               None   
4746  /annons/3916857  Lägenhet      Malmö                               None   
4790  /annons/3915511  Lägenhet      Malmö                      Valborgsgatan   
4803  /annons/3912771  Lägenhet      Malmö                           Brogatan   
4920  /annons/3860454  Lägenhet      Malmö                     Kungsörnsgatan   
4952  /annons/3918648  Lägenhet      Malmö                      Toftängsgatan   
4954  /annons/3854462  Lägenhet      Malmö                      Koggens gränd   
5034  /annons/3921244  Lägenhet   Västerås                       Bofinksvägen   
5113  /annons/3726671  Lägenhet      Malmö                               None   
5207  /annons/3924142  Lägenhet   Västerås        

We can see that all types except apartments have fairly high percentages of missing values. Let us assume that accomodations with null-valued housing societies will have the same housing society as the most frequent hosuing society for that zip code

In [35]:
#merge null-valued hs-rows with non-null valued hs on postal_code
df_accomodations_without_hs = df[(df["housing_society"].isnull())]
df_accomodations_with_hs = df[(df["housing_society"].notnull())]

df_merged = pd.merge(df_accomodations_without_hs,
                 df_accomodations_with_hs[['housing_society', 'postal_code']],
                 on='postal_code')

In [36]:
df_merged

id      type         city               street street_num  \
0     /bostad/1846982     Villa  Helsingborg       Rusthållsgatan         5A   
1     /bostad/1846982     Villa  Helsingborg       Rusthållsgatan         5A   
2     /bostad/1846982     Villa  Helsingborg       Rusthållsgatan         5A   
3     /bostad/1846982     Villa  Helsingborg       Rusthållsgatan         5A   
4     /bostad/2247046     Villa  Helsingborg          Grännagatan          7   
5     /bostad/2247046     Villa  Helsingborg          Grännagatan          7   
6     /bostad/2247046     Villa  Helsingborg          Grännagatan          7   
7     /bostad/2247046     Villa  Helsingborg          Grännagatan          7   
8     /bostad/2242476     Villa  Helsingborg              Alvägen         13   
9     /bostad/1882367     Villa  Helsingborg            Skolvägen         20   
10    /annons/3883320  Kedjehus  Helsingborg        Hittarpsvägen         15   
11    /annons/3877528     Villa  Helsingborg       Färjemansgatan         32   
12    /annons/3877528     Villa  Helsingborg       Färjemansgatan         32   
13    /annons/3877528     Villa  Helsingborg       Färjemansgatan         32   
14    /annons/3877528     Villa  Helsingborg       Färjemansgatan         32   
15    /annons/3877528     Villa  Helsingborg       Färjemansgatan         32   
16    /annons/3877528     Villa  Helsingborg       Färjemansgatan         32   
17    /annons/3877528     Villa  Helsingborg       Färjemansgatan         32   
18    /annons/3877528     Villa  Helsingborg       Färjemansgatan         32   
19    /annons/3877528     Villa  Helsingborg       Färjemansgatan         32   
20    /annons/3877528     Villa  Helsingborg       Färjemansgatan         32   
21    /annons/3877528     Villa  Helsingborg       Färjemansgatan         32   
22    /annons/3843289     Villa  Helsingborg    Johan Banérs gata         29   
23    /annons/3843289     Villa  Helsingborg    Johan Banérs gata         29   
24    /annons/3843289     Villa  Helsingborg    Johan Banérs gata         29   
25    /annons/3843289     Villa  Helsingborg    Johan Banérs gata         29   
26    /annons/3843289     Villa  Helsingborg    Johan Banérs gata         29   
27    /annons/3843289     Villa  Helsingborg    Johan Banérs gata         29   
28    /annons/3843289     Villa  Helsingborg    Johan Banérs gata         29   
29    /annons/3843289     Villa  Helsingborg    Johan Banérs gata         29   
...               ...       ...          ...                  ...        ...   
1829  /bostad/1959992     Villa    Stockholm  Tomtebobarnens gata         15   
1830  /annons/3895193     Villa    Stockholm      Lundagårdsvägen         45   
1831  /annons/3895193     Villa    Stockholm      Lundagårdsvägen         45   
1832  /bostad/1466962    Radhus    Stockholm       Alabastervägen         14   
1833  /bostad/1466962    Radhus    Stockholm       Alabastervägen         14   
1834  /bostad/1466962    Radhus    Stockholm       Alabastervägen         14   
1835  /bostad/1466962    Radhus    Stockholm       Alabastervägen         14   
1836  /annons/3906815    Radhus    Stockholm     Skebokvarnsvägen        196   
1837  /annons/3878328     Villa    Stockholm           Lindevägen         30   
1838  /annons/3826446     Villa    Stockholm    Kubikenborgsvägen         15   
1839  /annons/3826446     Villa    Stockholm    Kubikenborgsvägen         15   
1840  /annons/3826446     Villa    Stockholm    Kubikenborgsvägen         15   
1841  /annons/3815244     Villa    Stockholm      Rättar Vigs väg         26   
1842  /annons/3815244     Villa    Stockholm      Rättar Vigs väg         26   
1843  /annons/3915387     Villa    Stockholm       Båtsmanskroken         18   
1844  /annons/3915387     Villa    Stockholm       Båtsmanskroken         18   
1845  /annons/3928640    Radhus     Göteborg            Älvaleken          3   
1846  /bostad/2810487     Villa  Helsingborg          Kummelgatan        136   
1847  /bostad/21039

In [44]:
#Extract one of the most frequent housing societies for that zip-code
df_hs_mode = df_merged.groupby(df_merged.id).housing_society_y.agg(
    lambda x: pd.Series.mode(x)[0]).to_frame().reset_index().rename(columns={'housing_society_y':'housing_society'})

In [45]:
df_hs_mode.head(5)

id            housing_society
0  /annons/2336989  Riksbyggen BRF Blå Staden
1  /annons/2404586  Riksbyggen BRF Blå Staden
2  /annons/3037177  BRF Sverigehusets Kubik 3
3  /annons/3037486            BRF Flygledaren
4  /annons/3074547                BRF Kikaren

In [46]:
df.set_index('id', inplace=True)
df.update(df_hs_mode.set_index('id'))
df.reset_index(inplace=True)
df

id      type         city                street street_num  \
0     /annons/3848254  Lägenhet      Uppsala             Torngatan         24   
1      /bostad/280424  Lägenhet  Helsingborg      Lilla Bergaliden          1   
2      /bostad/285988  Lägenhet  Helsingborg            Skaragatan         27   
3     /bostad/3906240  Lägenhet  Helsingborg    Kristinelundsgatan         11   
4     /annons/3876994    Radhus  Helsingborg           Sommargatan         31   
5      /bostad/792134  Lägenhet  Helsingborg        Föreningsgatan         30   
6     /bostad/1846982     Villa  Helsingborg        Rusthållsgatan         5A   
7     /bostad/2247046     Villa  Helsingborg           Grännagatan          7   
8     /annons/3898060  Lägenhet  Helsingborg  Michael Löfmans gata         11   
9     /annons/3898172  Lägenhet  Helsingborg            Kullagatan         43   
10    /bostad/2242476     Villa  Helsingborg               Alvägen         13   
11     /bostad/281620  Lägenhet  Helsingborg           Koppargatan         4C   
12    /annons/3893570  Lägenhet      Uppsala      Väderkvarnsgatan        15D   
13    /bostad/3893571  Lägenhet     Västerås       Timmermansgatan          6   
14    /annons/3898019  Lägenhet  Helsingborg       Fågelsångsgatan         2B   
15    /annons/3872567  Lägenhet     Västerås         Rudbecksgatan         4B   
16     /bostad/184815  Lägenhet     Västerås            Knutsgatan         4C   
17    /annons/3892169    Radhus     Västerås            Floragatan        19D   
18    /annons/3892170  Lägenhet     Västerås            Syrenvägen         13   
19    /annons/3892352  Lägenhet     Västerås            Knotavägen         26   
20    /annons/3892774  Lägenhet     Västerås   Kristiansborgsallén         3A   
21    /annons/3893161  Lägenhet     Västerås        Johanneshöjden          2   
22     /bostad/198609  Lägenhet     Västerås           Stora gatan         2C   
23     /bostad/192778  Lägenhet     Västerås        Skjutbanegatan         3D   
24     /bostad/195378  Lägenhet     Västerås         Ringduvegatan         23   
25    /annons/3891423  Lägenhet  Helsingborg         Röamöllagatan         40   
26    /annons/3891431  Lägenhet  Helsingborg       Kungsfågelgatan         17   
27    /annons/3891806    Radhus  Helsingborg         Hagmarksgatan         1A   
28     /bostad/190801  Lägenhet     Västerås        Skjutbanegatan         4B   
29    /annons/3885658  Lägenhet     Västerås     Vetterslundsgatan         60   
...               ...       ...          ...                   ...        ...   
6881  /annons/3825179  Lägenhet    Stockholm            Atlasgatan         13   
6882  /annons/3929878  Lägenhet    Stockholm      Lm Ericssons väg         17   
6883  /bostad/4045803  Lägenhet    Stockholm        Torvsätravägen          7   
6884  /bostad/3086585     Villa     Göteborg              Hålvägen         22   
6885   /bostad/557047  Lägenhet     Göteborg         Godemansgatan         9C   
6886  /annons/3798133  Lägenhet     Göteborg           Befälsgatan          6   
6887   /bostad/523209  Lägenhet     Göteborg           Friggagatan         5B   
6888  /annons/3929173  Lägenhet     Göteborg        Olivedalsgatan        15B   
6889  /bostad/3938135  Lägenhet     Göteborg            Kungsgatan         7B   
6890   /bostad/528869  Lägenhet     Göteborg              Bomgatan         6B   
6891  /bostad/3945984  Lägenhet     Göteborg       Bratteråsbacken         35   
6892  /annons/3929207  Lägenhet     Göteborg        Kaggeledsgatan        38D   
6893  /annons/3912132  Lägenhet     Göteborg      Rustmästaregatan         15   
6894  /annons/3852884  Lägenhet     Göteborg          Batterigatan         8C   
6895  /annons/3922670  Lägenhet     Göteborg       Trädgårdsgärdet         66   
6896  /annons/3903073     Villa     Göteborg   Ella Hillbäcks gata          8   
6897  /annons/3835017  Lägenhet     Göteborg          Kobbeslätten         10   
6898  /bostad/3196718     Villa     Göteborg      

In [47]:
df[df.housing_society.isnull()]

id        type         city                 street  \
33    /annons/3887215       Villa     Västerås         Fältspatsvägen   
36    /bostad/2073765       Villa     Västerås           Malörtsvägen   
37    /annons/3891158      Parhus  Helsingborg               Rååvägen   
51    /bostad/2815284       Villa  Helsingborg              Långgatan   
52    /annons/3877132       Villa  Helsingborg              Lövstigen   
79    /bostad/1827908       Villa  Helsingborg        Prästkragegatan   
87    /bostad/1848983      Radhus     Västerås           Råbocksvägen   
89    /annons/3441795  Fritidshus     Västerås         Lönnbacksvägen   
91    /bostad/2090288       Villa     Västerås              Tallvägen   
103   /annons/3773854       Villa     Västerås          Sevalla Väsby   
110   /annons/3648513       Villa     Västerås               Östanbro   
111   /annons/3648523       Villa     Västerås               Östanbro   
116   /annons/3448556        Gård  Helsingborg     Södra Bokhagsvägen   
130   /bostad/2359982       Villa  Helsingborg            Kruareliden   
160   /annons/3352940    Kedjehus  Helsingborg           Hyacintgatan   
170   /bostad/1520486       Villa  Helsingborg             Briggvägen   
175   /bostad/1535383        Gård  Helsingborg        Bältebergavägen   
184   /bostad/2127857       Villa  Helsingborg         Jonstorpsvägen   
187   /bostad/3013086  Fritidshus     Västerås         Bergskogsvägen   
191   /annons/3826156       Villa     Västerås                Gottsta   
217   /bostad/1674775       Villa  Helsingborg         Jordugglegatan   
223   /annons/3873501       Villa  Helsingborg        Kvarnstensgatan   
242   /bostad/1771609      Parhus  Helsingborg          Missionsgatan   
258   /bostad/3016751       Villa     Västerås            Provargatan   
262   /bostad/1833715      Radhus     Västerås         Regnmätargatan   
275   /bostad/3050517       Villa     Västerås  Slagårda Skräddartorp   
276   /annons/3890680      Radhus     Västerås              Disas väg   
279   /bostad/1927501      Radhus     Västerås        Stångjärnsgatan   
301   /bostad/1927502      Radhus     Västerås        Stångjärnsgatan   
316   /bostad/2813070       Villa  Helsingborg           Rallaregatan   
...               ...         ...          ...                    ...   
6186  /bostad/2214103       Villa    Stockholm           Plommonvägen   
6279  /annons/3910038       Villa    Stockholm         Jultomtestigen   
6299  /annons/3897924       Villa    Stockholm       Stjärngossevägen   
6300  /annons/3909310       Villa    Stockholm         Hammerstavägen   
6303  /annons/3891803      Radhus    Stockholm  Staffan Stallares väg   
6309  /bostad/2196620       Villa    Stockholm        Norrskensbacken   
6334  /bostad/2205401       Villa    Stockholm         Sylvestergatan   
6343  /bostad/2216419      Radhus    Stockholm       Skattegårdsvägen   
6353  /bostad/2196710      Radhus    Stockholm           Snöripsvägen   
6425  /annons/3856637       Villa    Stockholm         Norregöksvägen   
6526  /bostad/2216434      Radhus    Stockholm       Skattegårdsvägen   
6544  /bostad/2197544       Villa    Stockholm             Kärvbacken   
6639  /annons/3816853       Villa    Stockholm           Sjöhagsvägen   
6725  /annons/3920594    Kedjehus    Stockholm           Vantörsvägen   
6746  /bostad/2206776      Radhus    Stockholm         Stävholmsgränd   
6783  /annons/3928130  Fritidshus      Uppsala              Husbyborg   
6794  /annons/3928513       Villa     Västerås              Nicksborg   
6795  /bostad/1876403       Villa     Västerås             Skerikesby   
6816  /bostad/3689228       Villa    Linköping            Vittengatan   
6821  /annons/3925634      Radhus    Linköping  Ivar Samuelssons gata   
6824  /annons/3675793    Lägenhet    Linköping           Gudruns gata   
6828  /bostad/3681984       Villa    Linköping          Furulidsvägen   
6830  /bostad/3693736       Villa    Linköping             Öjebyvägen   
6838

In [48]:
#num of null-values per type
df.housing_society.isnull().groupby(df["type"]).sum()

type
Fritidshus     86.0
Gård            9.0
Kedjehus       27.0
Lägenhet       35.0
Parhus         13.0
Radhus         51.0
Villa         292.0
Name: housing_society, dtype: float64

In [49]:
#percentage of null-values per type
df.housing_society.isnull().groupby(df["type"]).mean() * 100

type
Fritidshus     82.692308
Gård          100.000000
Kedjehus       67.500000
Lägenhet        0.584307
Parhus         24.528302
Radhus         26.701571
Villa          55.725191
Name: housing_society, dtype: float64

In [50]:
df.housing_society.isnull().sum()

513

We still have around 513 null values in *housing_society*. We could, for each *type*, check the most common housing society in that (1) district and (2) city and assign that value to the accomodation

In [51]:
df.groupby(['type', 'city', 'district'])['housing_society'].unique()

type        city         district                
Fritidshus  Göteborg     Askim                                                [None]
                         Fagerdahl                             [BRF Västra Nidingen]
                         Hisingen                              [BRF Västra Nidingen]
                         Hisings Kärra                         [BRF Västra Nidingen]
                         Kortedala                                            [None]
                         Lilleby                                              [None]
                         Linnarhult                                           [None]
                         Sillviks Havskoloni                                  [None]
                         Säve                                                 [None]
                         Torslanda Havsbadskoloni                  [BRF Flygledaren]
                         Västra Centrum                        [BRF Godhemsstjärnan]
               

# Longitude and Latitude

We know from earlier that there are no null-values in city, longitude and latitude. Knowing this, we can create a new feature which measures the distance between the accomodation and the city center using the haversine distance formula 

In [86]:
from geopy.geocoders import Nominatim
from geopy.extra.rate_limiter import RateLimiter

In [99]:
# extract all cities

cities = pd.DataFrame(df['city'].unique().tolist(), columns=['city'])
cities

city
0      Uppsala
1  Helsingborg
2     Västerås
3    Linköping
4       Örebro
5        Malmö
6    Stockholm
7     Göteborg

In [100]:
#Get longitude and latitude for each city
geolocator = Nominatim(user_agent='DataExploration')
geocode = RateLimiter(geolocator.geocode, min_delay_seconds=1)
cities['loc'] = cities['city'].apply(geocode)
cities['longitude'] = cities['loc'].apply(lambda loc: loc.longitude)
cities['latitude'] = cities['loc'].apply(lambda loc: loc.latitude)
cities.drop('loc', axis='columns', inplace=True)

In [101]:
cities

city  longitude   latitude
0      Uppsala  17.638744  59.858613
1  Helsingborg  12.704068  56.044198
2     Västerås  16.546368  59.611099
3    Linköping  15.624525  58.409813
4       Örebro  15.215118  59.274729
5        Malmö  13.000157  55.605293
6    Stockholm  18.071093  59.325117
7     Göteborg  11.967017  57.707233

In [ ]:
def haversine_array(lat1, lng1, lat2, lng2):
    lat1, lng1, lat2, lng2 = map(np.radians, (lat1, lng1, lat2, lng2))
    AVG_EARTH_RADIUS = 6371  # in km
    lat = lat2 - lat1
    lng = lng2 - lng1
    d = np.sin(lat * 0.5) ** 2 + np.cos(lat1) * np.cos(lat2) * np.sin(lng * 0.5) ** 2
    h = 2 * AVG_EARTH_RADIUS * np.arcsin(np.sqrt(d))
    return h

In [67]:
def get_distance_to_city_center(df_row):
    geolocator = Nominatim(user_agent='DataExploration')
    
    

In [71]:
get_distance_to_city_center(df.loc[0,:])

(12.7040684, 56.0441984)

In [21]:
con.close()